private/signer/v4/v4.go

```diff
@@ -326,7 +326,7 @@ func (v4 *signer) buildCanonicalHeaders(r rule, header http.Header) {
 			headerValues[i] = "host:" + v4.Request.URL.Host
 		} else {
 			headerValues[i] = k + ":" +
-				strings.Join(v4.Request.Header[http.CanonicalHeaderKey(k)], ",")
+				strings.Join(v4.signedHeaderVals[k], ",")
```

Comment:
It would seem that if some header that didn't satisfy any of the rules would pass through.

```go
if !r.IsValid(canonicalKey) {
 	continue // ignored header
}
```
Which means this header never gets added to the `v4.signedHeaderVals` and then would panic
```go
strings.Join(v4.signedHeaderVals[k], ",")
```
since it doesn't exist. I think a new method with less strict rules than `http.CanonicalHeaderKey` needs to be implemented.